escrow/src/main.sw

```diff
@@ -1,257 +1,355 @@
+    CreatedEscrowEvent,
+    DepositEvent,
+    DisputeEvent,
+    PaymentTakenEvent,
+    ProposedArbiterEvent,
+    ResolvedDisputeEvent,
+    ReturnedDepositEvent,
+    TransferredToSellerEvent,
+};
+
+use interface::Escrow;
 use std::{
-    address::Address,
     assert::require,
-    chain::auth::{AuthError, msg_sender},
-    context::{call_frames::{contract_id, msg_asset_id}, msg_amount, this_balance},
+    block::height,
+    chain::auth::msg_sender,
+    context::{call_frames::msg_asset_id, msg_amount},
     contract_id::ContractId,
     identity::Identity,
-    result::*,
-    revert::revert,
-    token::transfer_to_output,
+    logging::log,
+    option::Option,
+    result::Result,
+    storage::StorageMap,
+    token::transfer,
 };
 
-abi Escrow {
-    #[storage(read, write)]fn constructor(buyer: Address, seller: Address, asset: ContractId, asset_amount: u64) -> bool;
+storage {
+    /// Used as a temporary variable for containing a change, proposed by the seller, to the arbiter
+    /// Map(ID => Info)
+    arbiter_proposal: StorageMap<u64,
+    Option<Arbiter>> = StorageMap {
+    },
 
-    #[storage(read, write)]fn deposit() -> bool;
+    /// Information describing an escrow created via create_escrow()
+    /// Map(ID => Info)
+    escrows: StorageMap<u64,
+    EscrowInfo> = StorageMap {
+    },
 
-    #[storage(read, write)]fn approve() -> bool;
+    /// Number of created escrows
+    /// Used as an identifier for O(1) look-up in mappings
+    escrow_count: u64 = 0,
+}
 
-    #[storage(read, write)]fn withdraw() -> bool;
+impl Escrow for Contract {
+    #[storage(read, write)]fn accept_arbiter(identifier: u64) {
+        // The assertions ensure that only the buyer can accept a proposal if the escrow has not
+        // been completed and the seller has proposed a new arbiter
 
-    #[storage(read)]fn get_user_data(user: Address) -> (bool, bool);
+        let mut escrow = storage.escrows.get(identifier);
 
-    #[storage(read)]fn get_state() -> u64;
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(msg_sender().unwrap() == escrow.buyer.address, UserError::Unauthorized);
 
-    #[storage(read)]fn get_balance() -> u64;
-}
+        let arbiter = storage.arbiter_proposal.get(identifier);
 
-// TODO: add enums back in when they are supported in storage and "matching" them is implemented
-// enum State {
-//     Void: (),
-//     Pending: (),
-//     Completed: (),
-// }
-
-enum Error {
-    CannotReinitialize: (),
-    DepositRequired: (),
-    IncorrectAssetAmount: (),
-    IncorrectAssetId: (),
-    StateNotInitialized: (),
-    StateNotPending: (),
-    UnauthorizedUser: (),
-    UserHasAlreadyDeposited: (),
-}
+        // TODO: incomplete compiler defaults the Option<Arbiter> to not be None therefore fee check
+        // https://github.com/FuelLabs/sway/issues/2326
+        require(arbiter.is_some() && 0 < arbiter.unwrap().fee_amount, StateError::ArbiterHasNotBeenProposed);
```

Comment:
It is because the user can call this function before proposing has been set by the seller. This will allow the buyer to reset the arbiter to a zero'd data structure which may cause unexpected side effects.